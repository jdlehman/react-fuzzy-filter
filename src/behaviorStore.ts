import valoo from "valoo";

export interface Emitter<T> {
  (value: T): void;
  on: (fn: (value: T) => void) => valoo.Disposer;
}

export type Disposer = valoo.Disposer;

// emulates BehaviorSubject from rxjs
// (stores current state) and replays it on subscribe
export default function behaviorStore(): Emitter<string> {
  const store: valoo.Observable<string> = valoo("");
  let currentState: string;
  // save currentState before emitting
  const emit = function(value: string) {
    currentState = value;
    store(value);
  } as Emitter<string>;

  // emit currentState on subscribe
  // then use default valoo behavior
  emit.on = function(fn: (value: string) => void): valoo.Disposer {
    fn(currentState);
    return store.on(fn);
  };
  return emit;
}
