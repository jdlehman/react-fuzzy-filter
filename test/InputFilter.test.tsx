import React from "react";
import { fireEvent, render } from "react-testing-library";
import behaviorStore from "../src/behaviorStore";
import inputFilterFactory, { InputFilter } from "../src/InputFilter";

describe("InputFilter", () => {
  let Input: InputFilter;
  const store = behaviorStore("");
  beforeEach(() => {
    Input = inputFilterFactory(store);
  });

  describe("#render", () => {
    it("renders with defaults", () => {
      const utils = render(<Input />);
      expect(utils.container).toMatchSnapshot();
    });

    it("sets classPrefix", () => {
      const utils = render(<Input classPrefix="my-prefix" />);
      expect(utils.container).toMatchSnapshot();
    });

    it("sets inputProps", () => {
      const inputProps = { placeholder: "Search" };
      const utils = render(<Input inputProps={inputProps} />);
      const input = utils.getByPlaceholderText("Search");
      expect(input).toMatchSnapshot();
    });

    it("sets initialSearch", () => {
      const utils = render(<Input initialSearch="first search" />);
      const input = utils.getByValue("first search");
      expect(input).toMatchSnapshot();
    });
  });

  describe("#onChange", () => {
    let input: Element;
    const spy = jest.fn().mockImplementation((value: string) => value);
    beforeEach(() => {
      const utils = render(
        <Input onChange={spy} inputProps={{ placeholder: "Search" }} />
      );
      input = utils.getByPlaceholderText("Search");
    });

    it.skip("calls callback with search value", () => {
      fireEvent.change(input, {
        target: { value: "my string" },
      });
      expect(spy).toHaveBeenCalledWith("my string");
    });

    it("passes the value to the store", done => {
      const received = ["", "some input"];
      let ptr = 0;
      store.on(data => {
        expect(data).toEqual(received[ptr]);
        ptr++;
        if (ptr === 2) {
          done();
        }
      });

      fireEvent.change(input, {
        target: { value: "some input" },
      });
    });
  });
});
