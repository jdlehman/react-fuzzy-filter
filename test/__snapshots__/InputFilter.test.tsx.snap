// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFilter #render renders with defaults 1`] = `
<div>
  <input
    class="react-fuzzy-filter__input"
    value=""
  />
</div>
`;

exports[`InputFilter #render sets classPrefix 1`] = `
<div>
  <input
    class="my-prefix__input"
    value=""
  />
</div>
`;

exports[`InputFilter #render sets initialSearch 1`] = `
<input
  class="react-fuzzy-filter__input"
  value="first search"
/>
`;

exports[`InputFilter #render sets inputProps 1`] = `
<input
  class="react-fuzzy-filter__input"
  placeholder="Search"
  value=""
/>
`;
